### DISCUSSION TITLE
Troubleshooting the Disappearing 'category_id' Segment in Nested Category System

### DISCUSSION THREAD

**Discussion by aytackayin** [filamentphp/filament#9524] [Help]
I wanted to create a nested category system. Initially, I listed the categories using getEloquentQuery(). However, when I switched to the second page or changed the 'Per page' option, the request()->route('category_id') value was empty, and it displayed the root categories.
```
public static function getEloquentQuery(): Builder
{
	$categoryId = request()->route('category_id');
        if (!$categoryId) {
            $categoryId= 13;//default root category id
        }
        return parent::getEloquentQuery()->where('category_id', $category_id);
}
```
As a second approach, I set up a filter and applied filtering with ->default($categoryId) when the page was opened. This seemed to solve the issue, but I encountered the same problem with the $categoryId value when clicking the 'Create' button in CreateAction.

```
public static function table(Table $table): Table
    {
        $categoryId = request()->route('category_id');
        if (!$categoryId) {
            $categoryId= 13;//default root category id
        }
        return $table
            ->columns([
                TextColumn::make('title')
            ])
            ->recordUrl(fn (Category $record): string => static::getUrl('index', ['category_id' => $record]))
            ->headerActions([
                CreateAction::make()->url(route('filament.admin.resources.categories.create', ['category_id' => $categoryId])),
            ])

            ->filters([
                SelectFilter::make('category_id')
                    ->label('Parent Category')
                    ->native(false)
                    ->searchable()
                    ->options(fn (): array => Category::query()->pluck('title', 'id')->all())
                    ->default($categoryId),
            ]);
    }
    public static function getPages(): array
    {
        return [
            'index' => ListCategories::route('/{category_id?}'),
        ];
    }
```

In summary, in the places where I performed an action by extracting the category_id from the URL during reordering or changing pages in the table, the category_id segment was not visible. However, when I refreshed the page, the category_id value was present.

I want to understand the root cause of the issue and how to resolve it.

**Comment by aytackayin** [comment]
It seems I need to pass this 'category_id' to Livewire parameters, but I have no knowledge about Livewire. I've done some research, but couldn't find a solution. I would really appreciate your help if you could assist me with this.

**Comment by micraux** [comment]
Hi,

Same exact problem for me... :-(

My use case: create a new order from a client page.

I tried to create an action calling the order create page, passing the client_id as a parameter in the url.

This works well: I land on the right page with the client_id in the request.

But as soon as I click on a repeater field (products), the client_id disappears and I get an error message.

Does someone has any idea to solve this?

**Comment by micraux** [comment]
Hey, @aytackayin, I came to a working solution, at least for a create action...

```php
// In your CreateXxx.php (where Xxx is your resource name)
protected function beforeFill()
{
    if (isset($_REQUEST['category_id'])) {
        $this->data['category_id'] = $_REQUEST['category_id'];

        // If you need other columns (eg. the category name)
        $category = Category::find($_REQUEST['category_id']);
        $this->data['category_name'] = $category['name'];
    }
}
```

If you want to check in you form:
```php
Forms\Components\Placeholder::make('the_category_id') // CAUTION: do not use 'category_id'
    ->content(
        fn(Get $get) => $get('category_id')
    ),

Forms\Components\Placeholder::make('the_category_name') // CAUTION: do not use 'category_name'
    ->content(
        fn(Get $get) => $get('category_name')
    ),
```

Hope this helps...

**Comment by micraux** [comment]
Just found another (better) way, without using query strings...

In your Resource page, configure your route in order to include your parameter(s) in the url:

```php
// In your XxxResource.php  (where Xxx is your resource name)
public static function getPages(): array
{
    return [
        ...
        'create' => Pages\CreateXxx::route('/{category_id}/create'),
        ...
    ];
}
```
```php
// In your CreateXxx.php (where Xxx is your resource name)
public $category_id = null;

protected function beforeFill()
{
    $this->data['category_id'] = $this->category_id;
}
```
